### PR TITLE
CLI Local User Management

### PR DESCRIPTION
**Pull request by daviddob**
Added CLI and Testing Support for Local User Management

New commands added:
`shield create-user`
`shield edit-user`
`shield delete-user`
`shield users`
`shield user <uuid>`

### CODE REVIEW

File: cmd/shield/commands/dispatcher.go [hunk 1]
```diff
@@ -25,16 +25,18 @@ type HelpGroup struct {
...
-	JobsGroup     = &HelpGroup{name: "JOBS"}
-	ArchivesGroup = &HelpGroup{name: "ARCHIVES"}
-	AccessGroup   = &HelpGroup{name: "ACCESS"}
-	TasksGroup    = &HelpGroup{name: "TASKS"}
+	InfoGroup      = &HelpGroup{name: "INFO"}
+	MiscGroup      = &HelpGroup{name: "MISCELLANEOUS"}
+	BackendsGroup  = &HelpGroup{name: "BACKENDS"}
+	TargetsGroup   = &HelpGroup{name: "TARGETS"}
+	StoresGroup    = &HelpGroup{name: "STORES"}
+	SchedulesGroup = &HelpGroup{name: "SCHEDULES"}
```
Comment (jhunt) [inline cmd/shield/commands/dispatcher.go]: You should not be re-adding the SchedulesGroup -- maybe need a rebase against v8?

File: cmd/shield/main.go [hunk 2]
```diff
@@ -52,10 +53,14 @@ func main() {
 		After:     getopt.StringLong("after", 'A', "", "Only show archives that were taken after the given date, in YYYYMMDD format."),
 		Before:    getopt.StringLong("before", 'B', "", "Only show archives that were taken before the given date, in YYYYMMDD format."),
 		To:        getopt.StringLong("to", 0, "", "Restore the archive in question to a different target, specified by UUID"),
-		Limit:     getopt.StringLong("limit", 0, "", "Display only the X most recent tasks or archives"),
+		Limit:     getopt.StringLong("limit", 0, "", "Display only the X most recent tasks, archives, or users"),
 
 		Full: getopt.BoolLong("full", 0, "Show all backend information when listing backends"),
 
+		Backend: getopt.StringLong("backend", 'b', "", "Only show users with the specified backend."),
```
Comment (jhunt) [inline cmd/shield/main.go]: Where is this option used?  Looks to me like `shield users` only shows local users...

File: t/api [hunk 3]
```diff
@@ -502,8 +503,57 @@ try "Checking that various API endpoints require JSON payloads" # {{{
 
 # }}}
 
+# }}}
+#
+try "Checking that unimplemented v2HTTP methods return 501s" # {{{
```
Comment (jhunt) [inline t/api]: missing space between `v2` and `HTTP`

File: t/api [hunk 4]
```diff
@@ -502,8 +503,57 @@ try "Checking that various API endpoints require JSON payloads" # {{{
...
+   for type in auth/local/users; do
+     is "$(httpstat $method "http://$SHIELD_DAEMON_ADDR/v1/$type/sub/requests")" \
+        "501" "/v2/$type/sub/requests should HTTP 501 (not implemented) on a $method request"
+   done
+ done
+) 2>&1 | indent
+
+
+# }}}
+try "Checking that various v2API endpoints validate UUIDs" # {{{
```
Comment (jhunt) [inline t/api]: missing space between `v2` and `API`

File: t/api [hunk 5]
```diff
@@ -502,8 +503,57 @@ try "Checking that various API endpoints require JSON payloads" # {{{
...
+       is "$(httpstat $method "http://$SHIELD_DAEMON_ADDR/v1/$type/$uuid")" \
+          "501" "$method /v2/$type/$uuid should HTTP 501 (not implemented) on a malformed UUID"
+     done
+   done
+ done
+) 2>&1 | indent
+
+
+# }}}
+try "Checking that various v2API endpoints require JSON payloads" # {{{
```
Comment (jhunt) [inline t/api]: missing space between `v2` and `API`

File: t/api [hunk 6]
```diff
@@ -636,6 +686,26 @@ EOF
   "paused"    : false
 }
 EOF
+
+ create user exampleuser <<EOF
+{
+  "uuid":"355ccd3f-1d2f-49d5-937b-f4a12033a0cf",
+	"name":"exampleuser",
```
Comment (jhunt) [inline t/api]: spacing issue

File: t/api [hunk 7]
```diff
@@ -636,6 +686,26 @@ EOF
...
+  "account":"exampleuser",
+  "password":"foobar",
+  "sysrole":"technician"
+}
+EOF
+
+ create user admin <<EOF
+{
+  "uuid":"355ccd3f-1d2f-49d5-937b-f4a12033a0ce",
+	"name":"admin",
```
Comment (jhunt) [inline t/api]: spacing issue

File: t/api [hunk 8]
```diff
@@ -949,6 +1019,64 @@ try "Pausing and unpausing jobs" # {{{
...
+
+ run users --fuzzy a
+ context 'users with account "*a*"'
+ includes user exampleuser
+ includes user admin
+
+ run users --fuzzy m
+ context 'users with account "*m*"'
+ includes user exampleuser
+ includes user admin
```
Comment (jhunt) [inline t/api]: i would try `ex` instead of `m` and verify that we exclude admin, but include exampleuser

File: t/api [hunk 9]
```diff
@@ -949,6 +1019,64 @@ try "Pausing and unpausing jobs" # {{{
...
+
+ run users --sysrole technician
+ context 'users with sysrole technician'
+ includes user exampleuser
+ excludes user admin
+
+ run users --sysrole manager
+ context 'users with sysrole manager'
+ none 'no users with sysrole "manager"'
+
```
Comment (jhunt) [inline t/api]: We have no --backend tests -- is `--backend` being honored?

File: t/api [hunk 10]
```diff
@@ -1025,6 +1153,24 @@ EOF
   "paused"    : false
 }
 EOF
+
+update user admin <<EOF
+{
+  "uuid":"355ccd3f-1d2f-49d5-937b-f4a12033a0ce",
+	"name":"nobody",
```
Comment (jhunt) [inline t/api]: spacing issue

File: t/api [hunk 11]
```diff
@@ -1045,6 +1191,11 @@ try "Deleting things" # {{{
  can_delete target   redis-shared
  can_delete store    fs
  can_delete policy   short-term
+
+ context 'users can always be deleted'
+ can_delete user admin
+ can_delete user exampleuser
```
Comment (jhunt) [inline t/api]: Can we delete a non-existent user?

File: cmd/shield/main.go (hunk 2 above)
Comment (daviddob) [reply cmd/shield/main.go]: Yeah thats true, I had functionality to list users from all backends but hardcoded it to local as the request was for LUM, (not sure if we want to include that via a backend flag to show other than local)

File: t/api (hunk 9 above)
Comment (daviddob) [reply t/api]: see above comment

File: cmd/shield/main.go (hunk 2 above)
Comment (daviddob) [reply cmd/shield/main.go]: You also (shouldnt/wouldnt) be able to edit/delete/etc non local backend users as that is all handled via the mapping.

**Pull request merged**